keps/sig-node/2823-pod-addmission-plugin/README.md

```diff
@@ -0,0 +1,445 @@
+administrative rights on the Kubernetes cluster, API-layer constructs such as a
+Pod Security Policy or validation webhook may be deleted, which would
+effectively disable the effect of that PSP. Likewise, the second solution the
+team landed on — using Node taints and tolerations — was similarly bound to the
+Kubernetes API layer, which meant EKS users could modify those Node taints and
+tolerations, effectively disabling the effects. A third potential solution
+involving OCI hooks was then investigated. OCI hooks are separate executables
+that an OCI-compatible container runtime invokes that can modify the behaviour
+of the containers in a sandbox. While this solution would have solved the
+API-layer problem, it introduced other issues, such as the inefficiency of
+downloading the container image to the Node before the OCI hook was run. Since
+pods can be scheduled directly on the node using nodename parameter in pod spec,
+its required to add such validation in Kubelet itself.
+
+The final solution the EKS team settled on involved changing kubelet itself to
+support additional node-local Pod admission checks. This KEP outlines the EKS
+team’s approach and proposes upstreaming these changes to kubelet in order to
+allow extensible node-local last-minute validation checks. This functionality
+will enable cloud providers who support nodeless/serverless worker nodes to
+restrict Pods based on fields other than those already being validated by
+kubelet.
+
+Also, Kubernetes scheduler is extensible but isn’t omniscient about everything
+on the node hence there can be cases where a scheduling decision cannot be
+fulfilled by the node based on the current node conditions such as node-level
+resource allocation. The in-tree PodAdmitHandlers mentioned below are good
+examples for this.
+
+* Eviction manager rejects pods if the node is under high pressure (e.g., PIDs)
+* Topology manager rejects pods if the NUMA affinity cannot be satisfied
+
+There can be additional resource allocation conditions and policies handled
+externally to the kubelet, e.g.,
+
+* Alternative resource allocator implemented by the CRI which needs to reject
+  pods in conditions similar to the built-in resource managers.
+
+These can also benefit from a pluggable admission handler to reject pods before
+they are admitted by the kubelet. This is hard to be replicated by exposing more
+information to the scheduler which may be too costly and subject to race
+conditions.
+
+### Goals
+
+- Allow deployers of fully managed worker nodes to have control over Pods
+  running on those nodes.
+
+- Enable node-local Pod admission checks without requiring changes to kubelet.
+
+### Non-Goals
+
+- Move existing validations to “out of tree” plugins.
+
+- Change existing API-layer validation solutions such as Pod Security Policies
+  and validating admission webhooks.
+
+## Proposal
+
+The approach taken is similar to the container networking interface (CNI) plugin
```

Comment:
From a design perspective, CNI seems to have the most downsides of all node level plugin mechanims.  It's an RPC framework modulated via a process interface.  It's hard to extend, it has a narrow channel of error responses (limiting the amount of info that can be returned to users), and it complicates how security is handled (we have to defend against certain types of escalations that come due to process permission inheritance).

I would hold up CRI, CSI, and auth which leverage GRPC as more coherent, efficient, and evolvable interfaces.  I would expect a discussion in this kep about why binary shell execution was chosen over gRPC registration (in general we pick gRPC for most RPC callouts today, with the exception of webhook admission, which has a long history).  I would probably expect gRPC to be chosen as the "default" mechanism unless a very strong argument was made against it.

In long term, I consider it an open topic of discussion to move CNI to a gRPC style interface, and I would expect to raise the bar on components that don't either extend via the controller pattern or a gRPC pattern.  So it's good to start that discussion here.